File: frontend/src/types/automation.ts
/**
 * 定时任务（Automation）契约类型（镜像 backend/app/schemas/automation.py，CLAUDE.md 5.4）。
 * 改一端必须同步另一端。
 */
import type { ChartContent, TableContent } from '@/types/message'

export type AutomationAction = 'sql_report'

export interface AutomationChannelBinding {
  enabled: boolean
  channel_id: string | null
}

export interface AutomationNotification {
  on_success: AutomationChannelBinding
  on_failure: AutomationChannelBinding
}

/** 参数模板：SQL（含 ${var} 占位符）+ 用户变量默认值 + 可选图表配置 */
export interface AutomationParams {
  sql_text: string
  datasource_id: string | null
  variable_defaults?: Record<string, string>
  chart_config?: AutomationChartConfig | null
}

export interface AutomationChartConfig {
  chart_type: 'line' | 'bar' | 'pie' | 'scatter' | 'heatmap'
  dimension: string
  measures: string[]
  title?: string | null
}

export interface AutomationInfo {
  id: string
  name: string
  description: string | null
  action: string
  params: AutomationParams
  cron_expression: string
  timezone: string
  enabled: boolean
  notification: AutomationNotification | null
  last_run_at: string | null
  last_status: 'success' | 'failed' | 'running' | null
  next_run_at: string | null
  datasource_name: string | null
  /** cron 中文描述（后端 describe_cron 回填） */
  readable: string | null
  created_at: string | null
  updated_at: string | null
}

/** 单次运行记录（result 为 {table, chart?} 结果快照） */
export interface AutomationRunInfo {
  id: string
  automation_id: string
  status: 'running' | 'success' | 'failed'
  started_at: string | null
  finished_at: string | null
  duration_ms: number | null
  error: string | null
  params: Record<string, unknown> | null
  result: { table: TableContent; chart?: ChartContent } | null
}

/** parse 待确认草稿（不落库；cron 已由后端生成，datasource_id 已锁定） */
export interface AutomationDraft {
  name: string
  description: string | null
  params: AutomationParams
  cron_expression: string
  timezone: string
  notification: AutomationNotification | null
  readable: string | null
  datasource_name: string | null
}

File: backend/app/models/automation.py
"""自动化任务模型：泛化定时任务（动作 + 参数 + cron 调度 + 通知绑定）。

契约见 docs/定时任务与第三方通知方案.md 2.1 / docs/技术方案设计.md 2.2。
本期（Automation 骨架）只建 automations / automation_runs 两张表；
notification_channels / notification_logs 在通知模块落地。
"""
import datetime
import uuid

from sqlalchemy import Boolean, DateTime, ForeignKey, Index, Integer, String, func, text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Automation(Base):
    """自动化任务：cron 调度 + 动作参数 + 可选通知绑定。"""

    __tablename__ = "automations"
    __table_args__ = (
        Index("ix_automations_due", "enabled", "next_run_at"),
        Index("ix_automations_user_created", "user_id", "created_at"),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()")
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(String, nullable=True)  # 自然语言原句（自然语言创建时保留）
    action: Mapped[str] = mapped_column(
        String(30), nullable=False
    )  # sql_report（扩展位：export/webhook）
    # {datasource_id, sql_text, chart_config, variable_defaults}
    params: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    cron_expression: Mapped[str] = mapped_column(String(100), nullable=False)  # 5 段 cron
    timezone: Mapped[str] = mapped_column(String(50), nullable=False, default="Asia/Shanghai")
    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    # {on_success:{enabled,channel_id}, on_failure:{enabled,channel_id}}（通知接入后使用）
    notification: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    last_run_at: Mapped[datetime.datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    last_status: Mapped[str | None] = mapped_column(
        String(20), nullable=True
    )  # success/failed/running
    next_run_at: Mapped[datetime.datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    created_at: Mapped[datetime.datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime.datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    user = relationship("User", back_populates="automations")
    runs = relationship(
        "AutomationRun",
        back_populates="automation",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )


class AutomationRun(Base):
    """自动化任务运行历史：每次执行的结果快照（table + 可选 chart）。"""

    __tablename__ = "automation_runs"
    __table_args__ = (
        Index("ix_automation_runs_automation_started", "automation_id", "started_at"),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()")
    )
    automation_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("automations.id", ondelete="CASCADE"),
        nullable=False,
    )
    status: Mapped[str] = mapped_column(String(20), nullable=False)  # running/success/failed
    started_at: Mapped[datetime.datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    finished_at: Mapped[datetime.datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    duration_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    error: Mapped[str | None] = mapped_column(String, nullable=True)
    params: Mapped[dict | None] = mapped_column(JSONB, nullable=True)  # 本次注入后最终参数快照
    # {table: TableContent, chart?: ChartContent}
    result: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    created_at: Mapped[datetime.datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

    automation = relationship("Automation", back_populates="runs")
